src/jquery.ime.js

```diff
@@ -1,17 +1,41 @@
 ( function ( $ ) {
 	'use strict';
+	var TextEntryFactory, TextEntry, FormWidgetEntry, ContentEditableEntry,
+		defaultInputMethod;
 
 	// rangy is defined in the rangy library
 	/*global rangy */
 
 	/**
+	 * Minimal functionality from OOjs: Initialize a class for OO inheritance
+	 */
+	function initClass( fn ) {
+		fn.static = fn.static || {};
+	}
+
+	/**
+	 * Minimal functionality from OOjs: Initialize a class for OO inheritance
+	 * OOjs-like subclass initialization (minimal required functionality)
+	 */
+	function inheritClass( targetFn, originFn ) {
+		targetFn.parent = originFn;
+		targetFn.prototype = $.extend( {}, originFn.prototype );
+		targetFn.prototype.constructor = originFn.constructor;
+		targetFn.static = $.extend( {}, originFn.static );
+	}
+
+	/**
 	 * IME Class
+	 * @class
+	 *
+	 * @constructor
 	 * @param {Function} [options.helpHandler] Called for each input method row in the selector
 	 * @param {Object} options.helpHandler.imeSelector
 	 * @param {String} options.helpHandler.ime Id of the input method
 	 */
 	function IME( element, options ) {
 		this.$element = $( element );
+		this.textEntry = TextEntryFactory.static.singleton.wrap( this.$element );
```

Comment:
The TextEntryFactory.prototype.wrap method can return undefined. In that cases, dont we need a check for undefined somewhere in the initalization part to prevent errors while calling getTextBeforeSelection and replaceTextAtSelection?